Restructure index section

Hunterness opened the pull request:
This will hopefully remove some confusion for users around just showing off the command vs an example as well as removing examples for index providers when only one provider exists.

It also updates the links to not mention `administration` as the index chapter does not live in the administration chapter.

Hunterness committed:
Add discrete to sections it was missed on

The inner example sections don't need to be part of content navigation

Hunterness committed:
Restructure index examples

also remove setting index provider examples for indexes with only one provider

Hunterness committed:
Remove `administration` from index links

It was there since pre-4.3 when the index chapter was in the administration chapter, but as it isn't part of the new administration chapter it should probably go away

Hunterness committed:
Update table of content with new headers

also make them not discrete, so you can jump to the examples for the relevant index type easily

Hunterness committed:
Add intro text for new create example sections

neo-technology-commit-status-publisher commented:
Looks like you've updated the documentation!

Check out your changes at https://neo4j-docs-cypher-500.surge.sh
<!-- Sticky Pull Request Comment -->

Hunterness commented on modules/ROOT/pages/indexes-for-search-performance.adoc:
```diff
@@ -572,10 +535,9 @@ CREATE INDEX index_name FOR (n:Label) ON (n.prop1, ..., n.propN)
 ----
```
should we keep the command here or remove it as well?

Hunterness replied:
@stefano-ottolenghi do you have any opinions around these?

stefano-ottolenghi replied:
Ah, good point. I think we can get rid of those. There's the full syntax in two places, and plenty of examples for people to infer what the syntax for each particular case looks like.

Hunterness commented on modules/ROOT/pages/indexes-for-search-performance.adoc:
```diff
@@ -612,10 +573,9 @@ CREATE INDEX index_name FOR ()-[r:TYPE]-() ON (r.prop1, ..., r.propN)
 ----
```
should we keep the command here or remove it as well?

Hunterness commented on modules/ROOT/pages/indexes-for-search-performance.adoc:
```diff
@@ -393,94 +391,96 @@ If there are predicates on only a subset of the indexed properties, it will not
+
+
+[[indexes-create-range-index]]
+=== Creating a range index
```
should these new sections have discrete headers as well?

Hunterness replied:
Gonna go with no `discrete` on these

Hunterness commented on modules/ROOT/pages/indexes-for-search-performance.adoc:
```diff
@@ -639,79 +599,74 @@ Added 1 index.
+=== Creating a token lookup index
 
-A named relationship type lookup index for all relationships with any relationship type can be created with:
+Creating a token lookup index can be done with the `CREATE LOOKUP INDEX` command.
```
```suggestion
Creating a token lookup index (node label or relationship type lookup index) can be done with the `CREATE LOOKUP INDEX` command.
```

Hunterness commented on modules/ROOT/pages/indexes-for-search-performance.adoc:
```diff
@@ -639,79 +599,74 @@ Added 1 index.
-----
-CREATE LOOKUP INDEX index_name FOR ()-[r]-() ON EACH type(r)
-----
+Token lookup indexes (node label and relationship type lookup indexes) have only one index provider available, `token-lookup-1.0`, and no supported index configuration.
```
```suggestion
Token lookup indexes have only one index provider available, `token-lookup-1.0`, and no supported index configuration.
```

Hunterness commented on modules/ROOT/pages/indexes-for-search-performance.adoc:
```diff
@@ -957,11 +890,11 @@ Added 1 index.
-[[administration-indexes-create-a-point-index-specifying-both-the-index-provider-and-configuration]]
-=== Create a point index specifying both the index provider and configuration
+[[indexes-create-a-point-index-specifying-both-the-index-provider-and-configuration]]
+==== Create a point index specifying both the index provider and configuration
```
Should this example also be removed? since we only have one provider

stefano-ottolenghi replied:
I think so

Hunterness committed:
Apply suggestions from code review

stefano-ottolenghi reviewed (commented):
This looks much, much better! I really like the grouping under _Create index_, and the more considerate use of `[discrete]`.

Some thoughts:
1. I don't support renaming sections, which means breaking anchor links. If there's significant moving around of content that an old anchor doesn't make sense anymore, then fine, but if it's making it more descriptive or dropping `administration` ... I don't think there's enough gain given that the flipside is that links to those break. You have fixed a bunch in this manual, but who knows how many there are scattered around in docs.
1. I'd get rid of the large index that's at the root of `CREATE INDEX`. The main sections/groupings are found in the table of contents in the sidebar, while the individual sections have their own list (which I do like).
1. I would remove the `Syntax` section, break it out, and spread it to the relevant sections (maybe with a `Syntax` subsection when there's a lot, like in `CREATE INDEX`). Especially now with the broken-down show index groupings, we can put the relevant syntax in each group.
  Extra hints that syntax is in the wrong place are that 
  a) all admonitions found there are duplicated in `CREATE INDEX`
  b) `CREATE INDEX` opens by saying `Creating an index is done with the CREATE ... INDEX ... command`, which lacks so much detail to actually be helpful.
  c) it would also give more meaning to sentences like `Range indexes have only one index provider available, range-1.0, and no supported index configuration`, which are now a bit mysterious unless one has read the syntax section upfront.
1. Can we find a way to remove/merge/move at least some of the admonitions? I am looking in particular at `CREATE INDEX`, and having 2 regular (short) paragraphs and 4 admonitions is a sign that we can design it better. Furthermore, almost all examples have the remarks `The index name must be unique.` and many have `Note that the index is not immediately available, but is created in the background.`, which are duplicates of the opening admonitions.
1. Move `Composite index limitations` as last section - I doubt anybody cares about this _before_ reading some of the other sections.
1. I'm torn whether I support the `Result \n Added 1 index.` blocks in examples. I understand that they are there to show that the behavior is different if you append `IF NOT EXISTS`, but since
  a) there's really no variation, other than `Added 1 index` and `(no changes, no records)`
  b) there's a very descriptive admonition in the 20% of examples that exhibit the second result, saying `The index will not be created if there already exists an index with the same schema and type, same name or both.`
... I would take it away. Makes examples shorter.
1. Is the order of sections in `CREATE INDEX` the best? From my limited experience, I thought the best, sorted by popularity, would be `range > text > point > lookup` (since lookup are created by default on install).
2. It's a good time to get rid of all the `Query` titles on examples. Then if we also address 6., we can put the example title as title of the code block, remove the surrounding section `=====` bloat, and shrink the examples size by 50%!

Hunterness commented:
@stefano-ottolenghi 
1. the main reason that I removed the `administration` is that we now have a chapter called `Administration` which the indexes isn't included in so having administration in the link felt wrong to me. (your argument on link breaking is most likely why it wasn't changed back in 4.3 when the old administration chapter which included indexes was removed.) As for links, I tried searching both in the cypher manual, cheat sheet and operations manual for usages of the links I changed (didn't find any in the operations manual and have a PR up for the cheat sheet ones)
2. sure
3. When we did break the previous administration chapter we also restructured the constraint section into separate pages and it was talk about doing similarly for the index one, that however didn't happen. The constraints section has syntax and examples separated into two pages, which I kinda likes. It feels odd to me to not have all syntax parts collected (we do that for almost all of our teams command documentation, be that database management or indexes), however we quite often repeat the syntax pattern when it's more complicated in the start of the individual sections (for example one big table/section for user commands and then repeat of the create syntax in the create user section). That could maybe be an alternative?
4. `Note that the index is not immediately available, but is created in the background.` I thought I had removed those from each individual example and only had it at the start of the create (and syntax) section? maybe I missed some or remember wrong :shrug: I can take a look at combining them/make them into text
5. I think it has been there since before I started at neo XD but sure I can move it down :P
6. I think that if we want to update something like this it should be consistent with the constraints section, I'd probably do it separately and as part of this (especially since I'm unsure if I agree with removing them)
7. I have no preference to the order, I can switch it around
8. can you give an example for how it should look? I'm unsure I understand (and don't know enough asciidoc XD)

stefano-ottolenghi commented:
1. I had not checked other manuals, but great! Then totally ahead!
2. .
3. I see the structural _appeal_ that separating syntax from examples may have, but is it what you would _expect_ to find, if you were a person _unfamiliar_ with the topic? Cause I appreciate (and expect) much more a structure like this:
  ```
  Create index
  --/ Create range index
    --/ Syntax (only for range index)
    --/ Examples (only for range index)
--/ Create text index
    --/ Syntax (only for text index)
    --/ Examples (only for text index)
  ```
I am never going to be interested in the syntax of ALL commands _at once_ (unless I am a Cypher or field engineer looking for a quick reference, which we shouldn't take as intended readership).
In general, as a reader, if there is a section called `Create index`, I expect to find all information about index creation _in there_, and not that some bits are somewhere else. Does this resonate with you?

4. Same applies to the index uniqueness remarks :)

6+8. This is how I imagined it:

![Screenshot from 2023-04-11 10-19-13](https://user-images.githubusercontent.com/114478074/231100030-6ba16e5b-2c12-4d58-9390-690f926a30be.png)

or, if we don't care about the example title since there's already so much info surrounding it,

![Screenshot from 2023-04-11 10-20-50](https://user-images.githubusercontent.com/114478074/231100242-cc3cb3bc-42e1-45d6-a9b1-05a5b4e06c6f.png)

source is
```
==== Create a single-property range index for nodes

The example shows how to create a named range index on all nodes with label `Person` and having the `surname` property:

.Creating a named range index on a single property for all nodes with a particular label
[source, cypher]
----
CREATE INDEX node_range_index_name FOR (n:Person) ON (n.surname)
----

[NOTE]
====
The index name must be unique.
====
```
where `.Creating a named range index on a single property for all nodes with a particular label` was omitted in the second version.

I know the current docs style does not have padding and line underneath h3/h4 headers - I hacked it in the browser, but I'm gonna push it as a permanent css change.

Hunterness commented:
1. Is there other places I should check as well or is these 3 enough?
2. .
3. On one hand sure, on the other I like how the constraints section is structured and would probably rather do something like that and split it further? I'll make some more re-arranging and we can see if it's better
4. (edit) I checked and I have the comment on availability at the start of CREATE INDEX as well as start of each index section but not in each individual example. Is this good enough or do we want to remove it further?

6-8. I'll see what I can do
(edit): looking over your pictures again, I feel like it makes it seem less like examples? it no longer have the `example XX` text but are just random code? and if I add back syntax summaries in the sections I feel like there isn't really a big difference visually in what is a syntax code-block and what is an example? (more than the special characters in the text)

Hunterness commented:
To keep track of the updates:
- [x] 1
- [x] 2
- [x] 3
- [x] 4
- [x] 5
- [x] 6
- [x] 7
- [x] 8

Hunterness committed:
Review comments part 1

- move composite index limitations down
- remove setting index provider and config for point indexes
- remove large ToC in create index section (each subsection has their own smaller one)

Hunterness committed:
Review comments part 2

- switch order of index types in create index section: range, text, point, lookup
- merge some of the admonitions, or make them into regular text instead

Hunterness committed:
Review comments part 3

- list syntax summary of each command down in their sections (also keep the general syntax section as it's referenced to in places)

stefano-ottolenghi commented:
Great work!!

> Is there other places I should check as well or is these 3 enough?

I think operations manual was the most likely to have. It should be enough.

> On one hand sure, on the other I like how the constraints section is structured and would probably rather do something like that and split it further? I'll make some more re-arranging and we can see if it's better

As I said, we should think more about what inexperienced readers expect, what is best for them, and try to put ourselves in their shoes. The fact that a structure _looks nice_ does not mean that it's pedagogical. ... with that said, I see you have moved syntax inline in each section, so :+1: 

:arrow_right: The only change I'd recommend is breaking in two separate `source` blocks the two syntax options that are now currently in one single block :)

:arrow_right: Also, given the sentence `Range indexes have only one index provider available, range-1.0, and no supported index configuration.`, how about we take away the `configuration` bit from the corresponding syntax blocks, and also this sentence? As it is, it feels like we are showing something to users in the syntax, and then telling them: "btw, you can't do anything with that". The fact that the backend code _accepts_ that parameter doesn't mean that we should document it, if it's not of interest to users.

> I checked and I have the comment on availability at the start of CREATE INDEX as well as start of each index section but not in each individual example. Is this good enough or do we want to remove it further?

I find repetition noisy unless extremely necessary. If you deem if necessary then go for it, but I think we can assume that people will read the `CREATE INDEX` intro, so they can get that piece of info if they want. There's also not so many use cases where the <5 seconds before an index is ready are a game-changer, so I think we can bless readers with a sentence less for each section :D

> 6-8. I'll see what I can do

:+1: 

> looking over your pictures again, I feel like it makes it seem less like examples? it no longer have the example XX text but are just random code? and if I add back syntax summaries in the sections I feel like there isn't really a big difference visually in what is a syntax code-block and what is an example? (more than the special characters in the text)

There's not too much visual difference, true. To preserve it we can either 

- use `Syntax` as title of the syntax blocks, so that they get the blue bulky header, AND not give a title to examples, so they DON'T get the bulky header
- viceversa: no title to syntax blocks (as now), but title to all examples

:arrow_right: We could also make a `Syntax` section everywhere where relevant, `discrete` of course.

:arrow_right: One more think I can spot is that `Examples` is always just a paragraph in bold. I'd recommend it making a discrete subsection as well.

Hunterness committed:
Review comments part 4

- split the syntax blocks into two
- remove `options` from the indexes that ha no variation for it
- add notes about single provider in syntax section

Hunterness committed:
Review comments part 5

- update *Examples:* to discrete subsections

Hunterness committed:
Review comments part 6

- Fix index names to not conflict and update SHOW example results to reflect new names and index order

Hunterness commented:
> Also, given the sentence Range indexes have only one index provider available, range-1.0, and no supported index configuration., how about we take away the configuration bit from the corresponding syntax blocks, and also this sentence?

I think you mean the `options` part, and I can remove that on the indexes where it's not relevant. However I think the sentence should stay since the provider and (empty) config map gets returned in the show indexes and should therefore at least be mentioned in the docs somewhere.

> with that said, I see you have moved syntax inline in each section, so 👍

Yes, each section gets a syntax summary of relevant syntax parts, but I'm also keeping the syntax section for easy comparison and will leave the full syntax there.

Hunterness committed:
Remove missed reference when section was removed

Hunterness committed:
Review comments part 7

- Fix examples to not have example bubbles (or results when it's just statistics)

stefano-ottolenghi commented:
This is looking really good! My last thought is whether we should move the full syntax section to the bottom, either before or after the composite index remarks? If it is for reference/quick glance, as you said, I think it makes more sense that it comes after the more content-heavy sections that people will be interested into first.

Hunterness committed:
Review comments part 8

- Move syntax reference section down

stefano-ottolenghi commented:
I was reviewing the style changes I've done locally and I spotted a couple glitches:

- `Create a text index only if it does not already exist` is missing one equal sign in front
- `Only one node label lookup index can exist at the time.` admonitions: I think they'd be better placed _after_ each example, rather than before. Also, what does it mean, exactly? Is it a typo and should be `at a time` (as in _simultaneously)?
- Also, in case you missed it, https://github.com/neo4j/docs-cypher/pull/500#discussion_r1165121439

Hunterness committed:
Review comments part 9

- move notes about only one lookup index per token type
- remove syntax example for composite indexes
- fix missed = for one of the text index examples

stefano-ottolenghi reviewed (approved):
Awesome! 
Is there something else you'd like to change, or shall I merge?

Hunterness commented:
I think it's good to merge from my view👍

But didn't you have issues in the docs team with tests/TC not being run?

stefano-ottolenghi commented:
> But didn't you have issues in the docs team with tests/TC not being run?

Yours is one of the mysterious blessed PRs that got picked up correctly in TC (you can see the successful status of `pull-requests-cypher-test-dev-neo4j-dev` here).
